### DISCUSSION TITLE
Booting Windows 10 in a RAMdisk via IPXE possible (no physical drive in the computer)?

### DISCUSSION THREAD

**Discussion by ShinyMcShine1978** [ipxe/ipxe#565] [Q&A]
Hello *,

is it possible to use IPXE to load Windows 10 over the network and run it via a RAM drive on the machine?
Goals:
- no physical drives in the computer
- Windows is reverted back to an pre-defined state after a reboot
- no user data is saved on the computer
I am asked by my company to research this possibilty. Using the web, I found that there are the tools GRUB4DOS & FiraDisk RAMdrive. However, both tools are not developed anymore. And IPXE is much more appealing to me...

Is my envisages scenario possible with IPXE?

Thanks in advance for your answers to a newbie...

**Comment by mcb30** [comment]
You have two options:

1. Use iPXE and [`wimboot`](https://ipxe.org/wimboot) to boot into Windows 10 PE.  This is the closest you can get to running Windows from a RAM drive: it will load the whole system image into RAM at boot time.  However, you will find that many Windows applications will not run in Windows PE, and so this is unlikely to be what you want in practice.
2. Use iPXE to boot into Windows 10 from an iSCSI disk.  The disk image is hosted on your server, and is accessed on demand by the client.  This avoids the need to fit the whole system image into RAM, and lets you run a full version of Windows (instead of Windows PE).  By using copy-on-write snapshots on the iSCSI server (e.g. via LVM), you can get the "revert back to pre-defined state" behaviour that you want.

HTH,

Michael

**Reply by ShinyMcShine1978** [reply]
Hi Michael and thanks for the fast reply!

As you already guessed right, Windows 10 PE is not an option. RAM on the target machine is no problem, there's enough.
Regarding the second solution: The copy-on-write snapshots sound intriguing. If I'm not mistaken, my colleague already tried out something in that direction but did not get it working properly yet... I'll check with him.

The RAM solution is just something that our superiors would prefer. :) 

Thanks again!

**Reply by mcb30** [reply]
There have been some efforts towards getting "normal" Windows to run from a RAM disk, but you are likely to end up spending a lot of time dealing with obscure corner cases.  I would strongly recommend going down the iSCSI route.

Assuming that you use a standard Linux distro as the iSCSI target, then LVM thin snapshots will give you the copy-on-write behaviour, and targetcli will happily set up iSCSI targets for each snapshot.

**Reply by ShinyMcShine1978** [reply]
Thanks Michael, we'll look into it!

**Reply by sjevtic** [reply]
I have both approaches (1) and (2) working today on Win10 2004 exactly as @mcb30 describes.  To his point, WinPE has limited utility as a general purpose environment, so I only use it for setup and maintenance activities.  iSCSI boot through iPXE works flawlessly for (2) though, and I can revert to a snapshot on the iSCSI target whenever something goes wrong.

Windows 10 has a component called the Universal Write Filter that is a filesystem overlay driver that theoretically should enable use of a read-only root filesystem with the possibility for runtime changes to be redirected to RAM and discarded at shutdown:

https://docs.microsoft.com/en-us/windows-hardware/customize/enterprise/unified-write-filter

I never went down this path though as the ability to revert to snapshots on the iSCSI target was sufficient for my application so I have no idea how well this works in practice.

**Reply by ShinyMcShine1978** [reply]
Thanks for the feedback. Again, I will look into it. Got some other things on my plate right now though...

**Reply by admvir2oz** [reply]
> There have been some efforts towards getting "normal" Windows to run from a RAM disk, but you are likely to end up spending a lot of time dealing with obscure corner cases. I would strongly recommend going down the iSCSI route.
> 
> Assuming that you use a standard Linux distro as the iSCSI target, then LVM thin snapshots will give you the copy-on-write behaviour, and targetcli will happily set up iSCSI targets for each snapshot.

Hello. Previously I tried to work with loading Windows entirely using ipxe. First I install Windows on a physical disk, then copy the system image to the iscsi disk. Everything worked fine. But now I'm trying to do the same in UEFI mode. I try to load snponly.efi, it gets to loading Windows and I get an error (INACCESSIBLE BOOT DEVICE). As if something needs to be changed in Windows itself or in the ipxe script? The script is very simple - chain next server;
#!ipxe
set net0/gateway 0.0.0.0
set initiator-iqn iqn.GAMER20
sanboot --keep iscsi:xxx.xxx.xxx.xxx ::::iqn.orig2 || goto retry
:retry
reboot

**Reply by sjevtic** [reply]
I have not yet worked on on booting Windows on iSCSI using UEFI. Can you boot WinPE while sanhook'd and then install Windows from there?  I have found that this is generally the best approach for avoiding the inaccessible boot device error.

**Reply by tuxj** [reply]
Haven't used this method for a while, but for the record my 2¢ findings are:
- Installation using wimboot and regular WinPE installation usually goes well up until point, where system should start on itself after restart from installer and then it fails on inaccessible boot device, due to "missing iSCSI driver" (if I remember correctly, Windows by default doesn't have enabled iSCSI initiator during boot).
- My troubleshooting steps was using regular Windows installed to SSD, then enabled iSCSI service and then block device cloned to the iSCSI drive, it worked just fine.
- The major finding was, when I tried to install this from chainloaded WDS instead of my own boot.wim image (derived from regular ISO) with attached iSCSI drive, it worked itself on the first try and I think it is because WDS "cares" about enabling all needed drivers and services for the boot - is much more HW/boot environment aware.
- And if I remember correctly, I found out, this enablement of iSCSI during boot can be changed in the registry of installed OS from WinPE regedit (I might post it here sometime in future, if I will find exact steps...).

**Comment by shodanx2** [comment]
I remember back in like, 2000 something we had a winpe-like live OS and it had the task bar and a bunch of stuff. It felt like a full windows and it had many libraries, like the dot net stuff and more, "slipstreamed" into it. It felt like a full windows. 

Do we know how to prepare a winpe like this ?

**Reply by mawardiramliagmailcom** [reply]
Search "bartPE"
